### PR TITLE
daytona-bin: init at 0.9.0

### PR DESCRIPTION
**Pull request by drupol**
More info: https://github.com/daytonaio/daytona/

It is possible to build this package via the sources, however, the licence and trademark of the tool prevent anybody to re-use the name `Daytona`. Therefore, building a binary through Nix is forbidden, unless we name it differently (like `Iceweasel` vs `Firefox` see: https://en.wikipedia.org/wiki/Debian%E2%80%93Mozilla_trademark_dispute).

CC @Tpuljak @idagelic

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Tpuljak** [comment]
Hi @drupol,

Daytona Engineer here. 👋

Thanks for opening the PR!

We definitely have a plan to release a Nix package of the binary in the coming weeks.

If it's okay with you, we'd love to collaborate with you on this. Because the package will be rapidly changing in the next few weeks, we wouldn't want you to spend valuable time on packaging it before we believe it's ready.

You're welcome to join our [Community Slack](https://join.slack.com/t/daytonacommunity/shared_invite/zt-273yohksh-Q5YSB5V7tnQzX2RoTARr7Q) to discuss this further and chat about the upcoming features and fixes.

**Comment by drupol** [comment]
@Tpuljak I can't specify the `Version` while building the app.

I currently use `-X internal.Version=0.1.0` but that doesn't seems to have any effect. Do you have any clue?

**Comment by drupol** [comment]
> Hi @drupol,
> 
> Daytona Engineer here. 👋

Hi!

> 
> Thanks for opening the PR!
> 
> We definitely have a plan to release a Nix package of the binary in the coming weeks.

Cool!
 
> If it's okay with you, we'd love to collaborate with you on this. Because the package will be rapidly changing in the next few weeks, we wouldn't want you to spend valuable time on packaging it before we believe it's ready.

Sure, your help if definitely valuable for packaging it correctly.
 
> You're welcome to join our [Community Slack](https://join.slack.com/t/daytonacommunity/shared_invite/zt-273yohksh-Q5YSB5V7tnQzX2RoTARr7Q) to discuss this further and chat about the upcoming features and fixes.

I might probably join, but I rather keep our exchange public if it's ok for you. Like so if I'm not available for doing something, one of the other maintainers can also take the lead.

**Comment by drupol** [comment]
> @Tpuljak I can't specify the `Version` while building the app.
> 
> I currently use `-X internal.Version=0.1.0` but that doesn't seem to have any effect. Do you have any clue?

Please ignore my question, I found the answer in https://github.com/daytonaio/daytona/pull/86

**Comment by Tpuljak** [comment]
Hi @drupol,

Since we're new to packaging binaries for Nix, we'd love your input on how to approach this.

As stated in our License [here](https://github.com/daytonaio/daytona/tree/main#:~:text=We%20explicitly%20grant%20permission%20for%20you%20to%20make%20a%20build%20that%20includes%20our%20trademarks%20while%20developing%20Daytona%20itself.%20You%20may%20not%20publish%20or%20share%20the%20build%2C%20and%20you%20may%20not%20use%20that%20build%20to%20run%20Daytona%20for%20any%20other%20purpose.), the Daytona binary is built and distributed by the organization.

From your experience, does it make sense to include `daytonaio` in the maintainers entry of the package so that way it's clear that the organization is part of the packaging while you still maintain contribution rights?

We're talking with orgs with similar licenses to see the best way to approach this.

In the end, we'd love to give credit to you as well, of course, but we need to ensure that our license is covered.

**Comment by drupol** [comment]
Let's discuss this tomorrow, we can organise a meeting if you want.

**Comment by Tpuljak** [comment]
> Let's discuss this tomorrow, we can organize a meeting if you want.

Sure! We'll be in touch.

**Comment by drupol** [comment]
@Tpuljak What's the status with version 0.2.0? Do we need to have a meeting before merging this PR?

**Comment by mogery** [comment]
Heads up to those using both `daytona` and `programs.ssh` in `home-manager`: you'll need to add:
```
programs.ssh.includes = [ "daytona_config" ];
```
otherwise you'll be getting errors about a read-only filesystem when Daytona tries to add itself to the config.

**Comment by drupol** [comment]
@mogery Have you tested this derivation? Is it working for you?

**Comment by mogery** [comment]
> @mogery Have you tested this derivation? Is it working for you?

Yeah, I changed it to pull 0.5.0 but it's working flawlessly otherwise.

**Comment by drupol** [comment]
I just updated to version 0.5.0 and included the docker provider. However, it looks like daytona doesn't look in the patch. I might report that issue upstream so that the docker provider (and potentially others) will be correctly detected.

**Comment by mogery** [comment]
Oh, I haven't tried to make Docker work on NixOS. I don't actually use it like that on local, I just use the CLI on Nix with the profile of an Ubuntu VM I've set up that runs the Daytona server.

**Comment by drupol** [comment]
> Oh, I haven't tried to make Daytona + Docker work on NixOS. I don't actually use it like that on local, I just use the CLI on Nix with the profile of an Ubuntu VM I've set up that runs the Daytona server.

Do you mind making a demo or write a bit of documentation on how to use it?

> programs.ssh.includes = [ "daytona_config" ];

Also, where is coming from `daytona_config` in here?

**Comment by mogery** [comment]
> Do you mind making a demo or write a bit of documentation on how to use it?

Bit busy right now, but will do it ASAP.

> Also, where is coming from `daytona_config` in here?

[CLI creates it to set up SSH tunnels to the containers.](https://github.com/daytonaio/daytona/blob/56ba507e67226a092f251ca00cfd618c147b1326/cmd/daytona/config/ssh_file.go#L17-L67)

**Comment by drupol** [comment]
@mogery ^^

**Comment by drupol** [comment]
@mogery I'm looking for feedback before merging...

File: pkgs/by-name/da/daytona-bin/package.nix [hunk 1]
```diff
@@ -0,0 +1,68 @@
+{ stdenvNoCC
+, lib
+, fetchurl
+, callPackage
+, makeWrapper
+}:
+
+let
+  dayton-bin-provider-docker = callPackage ./providers/docker { };
```
Comment (TomaSajt) [inline pkgs/by-name/da/daytona-bin/package.nix:9]:
```suggestion
  daytona-bin-provider-docker = callPackage ./providers/docker { };
```

File: pkgs/by-name/da/daytona-bin/package.nix [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    fetchurl urls."${stdenvNoCC.hostPlatform.system}";
+
+  dontUnpack = true;
+
+  nativeBuildInputs = [
+    makeWrapper
+  ];
+
+  buildInputs = [
+    dayton-bin-provider-docker
```
Comment (TomaSajt) [inline pkgs/by-name/da/daytona-bin/package.nix:45]:
```suggestion
    daytona-bin-provider-docker
```

File: pkgs/by-name/da/daytona-bin/package.nix [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  ];
+
+  installPhase = ''
+    runHook preInstall
+    mkdir -p $out/bin $out/share/daytona/providers
+    cp -r $src $out/bin/daytona
+    chmod +x $out/bin/daytona
+    cp ${dayton-bin-provider-docker}/bin/daytona-docker-provider $out/share/daytona/providers/
```
Comment (TomaSajt) [inline pkgs/by-name/da/daytona-bin/package.nix:53]: IMO this is a a bit clearer.
```suggestion
    install -Dm755 $src $out/bin/daytona
    install -Dm755 ${daytona-bin-provider-docker}/bin/daytona-docker-provider -t $out/share/daytona/providers/
```

File: pkgs/by-name/da/daytona-bin/providers/docker/default.nix [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    in
+    fetchurl urls."${stdenvNoCC.hostPlatform.system}";
+
+  dontUnpack = true;
+
+  installPhase = ''
+    runHook preInstall
+    mkdir -p $out/bin
+    cp -r $src $out/bin/daytona-docker-provider
+    chmod +x $out/bin/daytona-docker-provider
```
Comment (TomaSajt) [inline pkgs/by-name/da/daytona-bin/providers/docker/default.nix:39]: IMO this is a bit clearer
```suggestion
    install -Dm755 $src $out/bin/daytona-docker-provider
```

**Review by TomaSajt — COMMENTED** [review]
Sorry, I can't actually test it either, though I have left some change recommendations

**Comment by TomaSajt** [comment]
I figured out how to run the server (it was pretty obvious), and it looks like it's trying search the wrong place for the providers.
After clearing `~/.config/daytona` (it was needed) and running the non-wrapped server, it downloaded the provider into `~/.config/daytona/providers/docker-provider/docker-provider`.
I can't really look into it more right now, but I think it's worth looking into.

I am not sure, which is better:
- prefetch the binaries into the nix-store
- let the program handle the fetching of the binaries

both work, because the binaries seem to be statically linked

**Comment by drupol** [comment]
> I figured out how to run the server (it was pretty obvious), and it looks like it's trying search the wrong place for the providers. After clearing `~/.config/daytona` (it was needed) and running the non-wrapped server, it downloaded the provider into `~/.config/daytona/providers/docker-provider/docker-provider`, so the `$out/share/daytona/providers` directory probably has to have the same nesting. I can't really look into it more right now, but I think it's worth looking into.
> 
> I am not sure, which is better:
> 
>     * prefetch the binaries into the nix-store
> 
>     * let the program handle the fetching of the binaries
> 
> 
> both work, because the binaries seem to be statically linked

I also encountered the same issue and I already had a chat with the Daytona devteam, they are going to have a look at it tomorrow. In the meantime, I've disabled the provider linkage until they fix the issue.

**Comment by TomaSajt** [comment]
It is easier and less diverging of upstream behaviour, so as long as it runs on NixOS by default, I think it's fine not to prefetch.

**Comment by drupol** [comment]
I removed the prefetch of docker provider. This is now ready for review.

**Comment by TomaSajt** [comment]
Seems like the wrapping was not actually removed

**Comment by drupol** [comment]
fixed!

**Comment by drupol** [comment]
@Tpuljak ready to be merged?

**Comment by Tpuljak** [comment]
Looks good to me.

Let's get approval from @metcalfc and @vedranjukic and we're good to go.

**Comment by Aleksanaa** [comment]
@Tpuljak  I want to clarify a few things.

In the NixOS/Nixpkgs ecosystem, maintainer refers to the person responsible for maintenance (tracking upstream updates and eliminating errors that occur in this ecosystem). Maintainers don't stand for upstream developers.

If your company's business operation model prompts you to maintain related packages on various distributions all by employees, you can ask employees to open a new PR and add them to the maintainer-list (or on behalf of your company's identity). But at the same time, I should note that Nixpkgs does not prevent anyone else who wants to maintain it from adding their name to the `meta.maintainers` field of the package.

In the case of maintenance by others, as long as your software functions properly on this platform, we may not consider the rest of your business plan as a blocker for any of our actions (unless this conflicts with laws and regulations). In other words, we may not wait for your permission for some actions. We may also not use your partnership model to communicate with you. Please tolerate Nixpkgs's operating mode, regards.

**Comment by Tpuljak** [comment]
@Aleksanaa thank you for the insights.

We are new to contributing packages to the Nix ecosystem and want to make sure we comply with its operating mode.

Because our license prohibits third parties from distributing the Daytona binary under the name `daytona`, we wanted to make sure all sides are in compliance with that. Additionally, we wanted to make sure @drupol is being properly attributed for his work in this PR.

We hope our way of communication is not misconstrued as us pushing our agenda against Nixpkgs's standards but as a way to ensure compliance with both parties. Daytona would love to be a part of the Nix ecosystem and accept best practices while doing so.

**Comment by metcalfc** [comment]
LGTM :shipit:

**Pull request merged**